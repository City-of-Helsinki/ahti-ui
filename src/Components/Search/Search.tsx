import React, { useEffect, useRef, useState } from 'react';
import styles from './Search.module.scss';
import { useTranslation } from 'react-i18next';
import mapData from '../../data/mapData.json';
import { escapeRegExp } from 'lodash';
import { useHistory } from 'react-router-dom';

import searchIcon from '../../assets/icons/search.svg';
import exitIcon from '../../assets/icons/cross.svg';

const features = mapData.features;

interface MapData {
  readonly name: string;
  readonly location: string;
  readonly type: string;
}

interface SearchItem extends MapData {
  readonly matchedPart: string;
}

const SearchItem: React.FC<SearchItem> = props => {
  const re = new RegExp(`(${escapeRegExp(props.matchedPart)})`, 'gi');
  const parts = props.name.split(re);

  let history = useHistory();

  return (
    <div
      className={styles.searchItem}
      onClick={() => history.push(`/map?name=${props.name}`)}
    >
      <img
        src={`/icons/type/${props.type}.svg`}
        alt="info image"
        style={{
          height: 30,
          width: 30,
          cursor: 'pointer',
        }}
      />
      <div className={styles.infoContainer}>
        <div className={styles.name}>
          {parts.map((part, i) =>
            re.test(part) ? (
              <mark key={i}>{part}</mark>
            ) : (
              <span key={i}>{part}</span>
            )
          )}
        </div>
        <div className={styles.address}>{props.location}</div>
      </div>
    </div>
  );
};

const Search: React.FC = () => {
  const [currentSearch, setCurrentSearch] = useState<string>('');
  const { i18n } = useTranslation();
  const history = useHistory();

  const inputRef = useRef<any>(null);
  useEffect(() => {
    if (inputRef) {
      inputRef.current.focus();
    }
  });

  const data: MapData[] = features.map((feature: any) => {
    const name = feature.properties[i18n.language].name;
    const location = feature.properties.address;
    const type = feature.properties.type;
    return { name, location, type };
  });

  return (
    <div className={styles.container}>
      <div className={styles.search}>
        <img src={exitIcon} alt="close" onClick={() => history.goBack()} />
        <div className={styles.searchInputWithIcon}>
          <input
            ref={inputRef}
            type="text"
            onChange={event => setCurrentSearch(event.target.value)}
            className={styles.searchInput}
          />
          <img src={searchIcon} alt="search" />
        </div>
      </div>
      {currentSearch === '' && (
        <>
          <p>{'Viimeisimm√§t'}</p>
          <p>{'Suosittelemme'}</p>
        </>
      )}
      {currentSearch !== '' &&
        data
          .filter((place: MapData) =>
            place.name.toLowerCase().includes(currentSearch.toLowerCase())
          )
          .map((place: MapData) => (
            <SearchItem {...place} matchedPart={currentSearch} />
          ))}
    </div>
  );
};

export default Search;
