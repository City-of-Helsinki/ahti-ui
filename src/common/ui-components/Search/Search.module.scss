@import '../../../styles/icons';

.container {
  width: 100%;
  position: absolute;
  background-color: white;
  padding-left: 1rem;
  padding-right: 1rem;

  /* Hide the scrollbar, but keep scrolling enabled */
  overflow-y: scroll;

  /* Firefox */
  scrollbar-width: none;

  /* Internet Explorer 10+ */
  -ms-overflow-style: none;
}

.container::-webkit-scrollbar {
  width: 0;
  height: 0;
}

.searchItem {
  display: flex;
  flex-direction: row;
}

.infoContainer {
  width: 100%;
  border-bottom: 1px;
  border-bottom-style: solid;
  border-bottom-color: var(--hds-ui-color-black-10);
  margin-left: 1rem;
}

.name {
  font-size: 16px;
  font-weight: 700;
  line-height: 26px;
}

.nameMark {
  all: unset;
  color: var(--hds-brand-color-coat-of-arms-blue);
}

.address {
  margin-bottom: 0.5rem;
  color: var(--hds-ui-color-black-70);
  font-weight: 400;
}

.searchInput {
  width: 100%;
  border: none;
  margin-left: 1rem;
}

.searchInput:focus {
  outline: none;
}

.searchInputWithIcon {
  display: flex;
  flex-direction: row;
  width: 100%;
  padding-bottom: 1rem;
  border-bottom: 1.5px;
  border-bottom-style: solid;
}

.infoImage {
  height: 30px;
  width: 30px;
  cursor: pointer;
}

.search {
  display: flex;
  flex-direction: row;
  margin-top: 1rem;
  margin-bottom: 1rem;
}
