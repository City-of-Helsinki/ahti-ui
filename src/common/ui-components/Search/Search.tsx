import React, { useEffect, useState } from 'react';
import styles from './Search.module.scss';
import { IconClose, IconSearch } from 'hds-react';
import {
  FeatureProperties,
  useFeaturesSearchQuery
} from '../../../domain/api/generated/types.d';
import { IconLocation } from 'hds-react';
import escapeRegExp from './utils/escapeRegExp';
import classNames from 'classnames';

export interface SearchData {
  readonly id: string;
  readonly name: string;
  readonly location?: string;
  readonly category?: string;
}

interface CategoryIconProps {
  readonly category?: string;
}

const CategoryIcon: React.FC<CategoryIconProps> = ({ category }) => {
  // BE only has ahti:category:island for the time being
  // other categories are null.
  if (category === 'ahti:category:island') {
    return <IconLocation className={styles.bigIcon} />;
  } else {
    return <IconLocation className={styles.bigIcon} />;
  }
};

interface SearchItemProps extends SearchData {
  readonly currentSearch: string;
  onSelect(id: string): void;
}

export const SearchItem: React.FC<SearchItemProps> = ({
  id,
  name,
  location,
  category,
  currentSearch,
  onSelect
}) => {
  const re = new RegExp(`(${escapeRegExp(currentSearch)})`, 'gi');
  const parts = name.split(re);

  return (
    <div className={styles.searchItem} onClick={() => onSelect(id)}>
      <CategoryIcon category={category} />
      <div className={styles.infoContainer}>
        <div className={styles.name}>
          {parts.map((part: string, id: number) =>
            re.test(part) ? (
              <mark key={id} className={styles.nameMark}>
                {part}
              </mark>
            ) : (
              <span key={id}>{part}</span>
            )
          )}
        </div>
        <div className={styles.address}>{location}</div>
      </div>
    </div>
  );
};

interface SearchProps {
  readonly className?: string;
  readonly maxItems?: number;
  onSelect(id: string): void;
}

const Search: React.FC<SearchProps> = ({
  className,
  maxItems = 20,
  onSelect
}) => {
  const { data } = useFeaturesSearchQuery();
  const [searchResults, setSearchResults] = useState<SearchData[]>([]);
  const [currentSearch, setCurrentSearch] = useState<string>('');

  useEffect(() => {
    if (data && data.features && currentSearch !== '') {
      const properties = data.features.edges.reduce((acc, edge) => {
        if (edge && edge.node && edge.node.properties) {
          return [...acc, edge.node.properties];
        }
        return acc;
      }, []);
      setSearchResults(
        properties
          .map((property: FeatureProperties) => {
            return {
              id: property.ahtiId,
              name: property.name,
              location: property?.contactInfo?.address?.municipality,
              category: property.category?.id
            };
          })
          .filter((searchResult: SearchData) =>
            searchResult.name
              .toLowerCase()
              .includes(currentSearch.toLowerCase())
          )
          .slice(0, maxItems)
      );
    } else {
      setSearchResults([]);
    }
  }, [currentSearch, data]);

  return (
    <div className={classNames(styles.container, className)}>
      <div className={styles.search}>
        <div className={styles.searchInputWithIcon}>
          <IconSearch className={styles.bigIcon} />
          <input
            type="text"
            value={currentSearch}
            className={styles.searchInput}
            onChange={event => setCurrentSearch(event.target.value)}
          />
          <button aria-label="close" onClick={() => setCurrentSearch('')}>
            <IconClose className={styles.bigIcon} />
          </button>
        </div>
      </div>
      {searchResults.map((item: SearchData, id: number) => (
        <SearchItem
          key={id}
          {...item}
          currentSearch={currentSearch}
          onSelect={onSelect}
        />
      ))}
    </div>
  );
};

export default Search;
