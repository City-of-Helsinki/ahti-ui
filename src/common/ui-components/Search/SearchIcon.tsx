import React from 'react';
import classNames from 'classnames';

import styles from './Search.module.scss';

const SearchIcon: React.FC<any> = ({}) => {
  return (
    <div className={classNames(styles.searchIcon)}>
      <span></span>
      <span></span>
      <span></span>
      <span></span>
    </div>
  );
};

export default SearchIcon;
