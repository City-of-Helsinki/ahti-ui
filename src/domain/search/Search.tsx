import React, { useEffect, useRef, useState } from 'react';
import styles from './Search.module.scss';
import { useTranslation } from 'react-i18next';
import mapData from '../../data/mapData.json';
import { escapeRegExp } from 'lodash';
import { useHistory } from 'react-router-dom';

import searchIcon from '../../assets/icons/search.svg';
import exitIcon from '../../assets/icons/cross.svg';

const features = mapData.features;

export interface SearchData {
  readonly name: string;
  readonly location: string;
  readonly type: string;
}

interface SearchItemProps extends SearchData {
  readonly matchedPart: string;
  onSelect(item: string): void;
}

export const SearchItem: React.FC<SearchItemProps> = props => {
  const re = new RegExp(`(${escapeRegExp(props.matchedPart)})`, 'gi');
  const parts = props.name.split(re);

  return (
    <div
      className={styles.searchItem}
      onClick={() => props.onSelect(props.name)}
    >
      <img
        src={`/icons/type/${props.type}.svg`}
        alt="info image"
        className={styles.infoImage}
      />
      <div className={styles.infoContainer}>
        <div className={styles.name}>
          {parts.map((part, i) =>
            re.test(part) ? (
              <mark key={i}>{part}</mark>
            ) : (
              <span key={i}>{part}</span>
            )
          )}
        </div>
        <div className={styles.address}>{props.location}</div>
      </div>
    </div>
  );
};

interface SearchProps {
  readonly data: SearchData[];
  onClose(): void;
  onSelect(item: string): void;
}

const Search: React.FC<SearchProps> = ({ data, onClose, onSelect }) => {
  const [currentSearch, setCurrentSearch] = useState<string>('');
  const inputRef = useRef<any>(null);

  useEffect(() => {
    if (inputRef) {
      inputRef.current.focus();
    }
  });

  return (
    <div className={styles.container}>
      <div className={styles.search}>
        <img src={exitIcon} alt="close" onClick={() => onClose()} />
        <div className={styles.searchInputWithIcon}>
          <input
            ref={inputRef}
            type="text"
            onChange={event => setCurrentSearch(event.target.value)}
            className={styles.searchInput}
          />
          <img src={searchIcon} alt="search" />
        </div>
      </div>
      {currentSearch === '' && (
        <>
          <p>{'Viimeisimm√§t'}</p>
          <p>{'Suosittelemme'}</p>
        </>
      )}
      {currentSearch !== '' &&
        data
          .filter((item: SearchData) =>
            item.name.toLowerCase().includes(currentSearch.toLowerCase())
          )
          .map((item: SearchData, id: number) => (
            <SearchItem
              key={id}
              {...item}
              matchedPart={currentSearch}
              onSelect={onSelect}
            />
          ))}
    </div>
  );
};

export const SearchWithHistoryHook: React.FC = () => {
  const history = useHistory();
  const { i18n } = useTranslation();

  const data: SearchData[] = features.map((feature: any) => {
    const name = feature.properties[i18n.language].name;
    const location = feature.properties.address;
    const type = feature.properties.type;
    return { name, location, type };
  });
  return (
    <Search
      data={data}
      onClose={() => history.goBack()}
      onSelect={name => history.push(`/map?name=${name}`)}
    />
  );
};

export default Search;
