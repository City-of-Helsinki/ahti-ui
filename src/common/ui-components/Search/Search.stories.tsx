import React from 'react';
import Search from './Search';
import { ApolloProvider } from '@apollo/react-common';
import graphQLClient from '../../../domain/api';

export default {
  title: 'Search',
  component: Search,
  decorators: [
    (storyFn: () => React.ReactNode) => (
      <ApolloProvider client={graphQLClient}>{storyFn()}</ApolloProvider>
    )
  ]
};

export const DefaultSearch = () => (
  <Search
    onSelect={item => {
      alert(`clicked: ${item}`);
    }}
  />
);

DefaultSearch.story = {
  name: 'Default Search'
};
