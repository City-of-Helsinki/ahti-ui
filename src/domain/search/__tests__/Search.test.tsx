import React from 'react';
import Search, { SearchData, SearchItem } from '../Search';
import { shallow, ShallowWrapper } from 'enzyme';

const mockData: SearchData[] = [
  { name: 'Venepaikka', location: 'Helsinki', type: 'boat' },
];

describe('Search', () => {
  let searchWrapper: ShallowWrapper;
  let mockOnClose: jest.Mock;

  beforeEach(() => {
    mockOnClose = jest.fn();
    searchWrapper = shallow(
      <Search data={mockData} onClose={mockOnClose} onSelect={jest.fn()} />
    );
  });

  it('matches snapshot', () => {
    expect(searchWrapper.html()).toMatchSnapshot();
  });

  it('returns items on search', () => {
    expect(searchWrapper.find(SearchItem)).toHaveLength(0);
    searchWrapper.find('input').simulate('change', { target: { value: 'a' } });
    expect(searchWrapper.find(SearchItem)).toHaveLength(1);
  });

  it('calls close function on click', () => {
    expect(mockOnClose.mock.calls.length).toBe(0);
    searchWrapper.find('img[alt="close"]').simulate('click');
    expect(mockOnClose.mock.calls.length).toBe(1);
  });
});

describe('SearchItem', () => {
  let searchItemWrapper: ShallowWrapper;
  let mockOnSelect: jest.Mock;

  beforeEach(() => {
    mockOnSelect = jest.fn();
    searchItemWrapper = shallow(
      <SearchItem
        matchedPart={'Vene'}
        onSelect={mockOnSelect}
        {...mockData[0]}
      />
    );
  });

  it('matches snapshot', () => {
    expect(searchItemWrapper.html()).toMatchSnapshot();
  });

  it('calls onSelect on item click', () => {
    expect(mockOnSelect.mock.calls.length).toBe(0);
    searchItemWrapper.simulate('click');
    expect(mockOnSelect.mock.calls.length).toBe(1);
  });
});
