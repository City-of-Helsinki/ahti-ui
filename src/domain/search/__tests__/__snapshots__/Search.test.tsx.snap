// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search matches snapshot 1`] = `"<div class=\\"container\\"><div class=\\"search\\"><img src=\\"cross.svg\\" alt=\\"close\\"/><div class=\\"searchInputWithIcon\\"><input type=\\"text\\" class=\\"searchInput\\"/><img src=\\"search.svg\\" alt=\\"search\\"/></div></div><p>Latest</p><p>We recommend</p></div>"`;

exports[`SearchItem matches snapshot 1`] = `"<div class=\\"searchItem\\"><img src=\\"/icons/type/boat.svg\\" alt=\\"info icon\\" class=\\"infoImage\\"/><div class=\\"infoContainer\\"><div class=\\"name\\"><span></span><mark>Vene</mark><span>paikka</span></div><div class=\\"address\\">Helsinki</div></div></div>"`;
